File: webapp/pdb_analysis.py
# -*- coding: utf-8 -*-
"""Modulo de analise estrutural de proteinas (formato PDB).

Reune, em funcoes puras que operam sobre o TEXTO de um arquivo PDB, a logica
dos Programas 01 a 05 do trabalho original:

  - separar_cadeias        (Programas 01/02)
  - estatisticas_cadeias   (Programa 03)
  - frequencia_aminoacidos (Programa 04)
  - buscar_motivo          (Programa 05)

As funcoes nao fazem I/O de arquivos: recebem texto e devolvem estruturas
Python (dicts/listas), prontas para serem expostas por uma API web.
"""

from __future__ import annotations

# 20 aminoacidos padrao e a conversao de 3 letras para 1 letra
AMINOACIDOS_PADRAO = [
    "ALA", "ARG", "ASN", "ASP", "CYS", "GLN", "GLU", "GLY", "HIS", "ILE",
    "LEU", "LYS", "MET", "PHE", "PRO", "SER", "THR", "TRP", "TYR", "VAL",
]

TRES_PARA_UMA = {
    "ALA": "A", "ARG": "R", "ASN": "N", "ASP": "D", "CYS": "C",
    "GLN": "Q", "GLU": "E", "GLY": "G", "HIS": "H", "ILE": "I",
    "LEU": "L", "LYS": "K", "MET": "M", "PHE": "F", "PRO": "P",
    "SER": "S", "THR": "T", "TRP": "W", "TYR": "Y", "VAL": "V",
}


def _linhas_coordenada(texto_pdb: str):
    """Itera apenas sobre as linhas de coordenada relevantes (ATOM)."""
    for linha in texto_pdb.splitlines():
        if linha.startswith("ATOM"):
            yield linha


def listar_ids_cadeias(texto_pdb: str) -> list[str]:
    """Retorna os identificadores de cadeia presentes, em ordem de aparicao."""
    vistos: list[str] = []
    for linha in texto_pdb.splitlines():
        registro = linha[0:6].strip()
        if registro in ("ATOM", "HETATM", "TER") and len(linha) >= 22:
            cid = linha[21]
            if cid == " ":
                cid = "_"
            if cid not in vistos:
                vistos.append(cid)
    return vistos


def separar_cadeias(texto_pdb: str) -> dict[str, str]:
    """Separa o PDB por cadeia. Retorna {id_cadeia: texto_pdb_da_cadeia}.

    Inclui ATOM/HETATM/TER/ANISOU e finaliza cada cadeia com END.
    """
    cadeias: dict[str, list[str]] = {}
    for linha in texto_pdb.splitlines():
        registro = linha[0:6].strip()
        if registro in ("ATOM", "HETATM", "TER", "ANISOU") and len(linha) >= 22:
            cid = linha[21]
            if cid == " ":
                cid = "_"
            cadeias.setdefault(cid, []).append(linha)

    resultado: dict[str, str] = {}
    for cid in sorted(cadeias):
        resultado[cid] = "\n".join(cadeias[cid]) + "\nEND\n"
    return resultado


def _residuos_ordenados(texto_pdb: str, id_cadeia: str | None = None):
    """Lista (nome_residuo, num_residuo) de uma cadeia, na ordem, sem repetir."""
    sequencia = []
    vistos = set()
    for linha in _linhas_coordenada(texto_pdb):
        if len(linha) < 27:
            continue
        cid = linha[21]
        if cid == " ":
            cid = "_"
        if id_cadeia is not None and cid != id_cadeia:
            continue
        nome = linha[17:20].strip()
        num = linha[22:27].strip()
        chave = (cid, nome, num)
        if chave not in vistos:
            vistos.add(chave)
            sequencia.append((nome, num))
    return sequencia


def estatisticas_cadeias(texto_pdb: str) -> dict:
    """Estatisticas por cadeia: atomos, residuos, aminoacidos diferentes.

    Retorna {"cadeias": [...], "maior": {...}, "menor": {...}}.
    """
    por_cadeia: dict[str, dict] = {}

    for linha in _linhas_coordenada(texto_pdb):
        if len(linha) < 27:
            continue
        cid = linha[21]
        if cid == " ":
            cid = "_"
        nome = linha[17:20].strip()
        num = linha[22:27].strip()

        d = por_cadeia.setdefault(cid, {"atomos": 0, "residuos": set(), "aa": set()})
        d["atomos"] += 1
        d["residuos"].add(num)
        if nome in AMINOACIDOS_PADRAO:
            d["aa"].add(nome)

    cadeias = []
    for cid in sorted(por_cadeia):
        d = por_cadeia[cid]
        cadeias.append({
            "cadeia": cid,
            "atomos": d["atomos"],
            "residuos": len(d["residuos"]),
            "aminoacidos_diferentes": len(d["aa"]),
        })

    resposta = {"cadeias": cadeias, "maior": None, "menor": None}
    if cadeias:
        resposta["maior"] = max(cadeias, key=lambda c: c["residuos"])
        resposta["menor"] = min(cadeias, key=lambda c: c["residuos"])
    return resposta


def frequencia_aminoacidos(texto_pdb: str, id_cadeia: str | None = None) -> dict:
    """Frequencia e porcentagem dos 20 aminoacidos (de uma cadeia ou de todas).

    Retorna {"total": int, "itens": [{"aa","contagem","porcentagem"}, ...]}.
    """
    contagem = {aa: 0 for aa in AMINOACIDOS_PADRAO}
    for nome, _num in _residuos_ordenados(texto_pdb, id_cadeia):
        if nome in contagem:
            contagem[nome] += 1

    total = sum(contagem.values())
    itens = []
    for aa in AMINOACIDOS_PADRAO:
        pct = (100.0 * contagem[aa] / total) if total else 0.0
        itens.append({"aa": aa, "contagem": contagem[aa], "porcentagem": round(pct, 2)})

    return {"total": total, "itens": itens}


def sequencia_uma_letra(texto_pdb: str, id_cadeia: str) -> str:
    """Sequencia de uma letra de uma cadeia (X para residuo nao-padrao)."""
    letras = []
    for nome, _num in _residuos_ordenados(texto_pdb, id_cadeia):
        letras.append(TRES_PARA_UMA.get(nome, "X"))
    return "".join(letras)


def buscar_motivo(texto_pdb: str, motivo: list[str]) -> list[dict]:
    """Procura o motivo (lista de aminoacidos de 3 letras) em todas as cadeias.

    Retorna lista de ocorrencias: {"cadeia","posicao","residuos"}.
    posicao = numero do residuo (do PDB) onde a ocorrencia comeca.
    """
    motivo = [m.strip().upper() for m in motivo if m.strip()]
    ocorrencias: list[dict] = []
    if not motivo:
        return ocorrencias

    n = len(motivo)
    for cid in listar_ids_cadeias(texto_pdb):
        seq = _residuos_ordenados(texto_pdb, cid)
        nomes = [r[0] for r in seq]
        for i in range(len(nomes) - n + 1):
            if nomes[i:i + n] == motivo:
                nums = [seq[i + k][1] for k in range(n)]
                ocorrencias.append({
                    "cadeia": cid,
                    "posicao": seq[i][1],
                    "residuos": nums,
                })
    return ocorrencias


def analisar_completo(texto_pdb: str) -> dict:
    """Roda todas as analises de uma vez (usado pelo endpoint principal)."""
    cadeias_ids = listar_ids_cadeias(texto_pdb)
    stats = estatisticas_cadeias(texto_pdb)

    freq_por_cadeia = {}
    sequencias = {}
    for cid in cadeias_ids:
        freq_por_cadeia[cid] = frequencia_aminoacidos(texto_pdb, cid)
        sequencias[cid] = sequencia_uma_letra(texto_pdb, cid)

    return {
        "cadeias": cadeias_ids,
        "estatisticas": stats,
        "frequencias": freq_por_cadeia,
        "sequencias": sequencias,
    }

File: docs/pdb-analysis.js
// ProtAnalysis - logica de analise estrutural de proteinas (formato PDB).
// Versao JavaScript (roda 100% no navegador) das funcoes do modulo Python
// pdb_analysis.py: cadeias, estatisticas, frequencia de aminoacidos, sequencia
// e busca de motivos. Os indices de coluna seguem o formato PDB.

"use strict";

const ProtAnalysis = (() => {

  const AMINOACIDOS_PADRAO = [
    "ALA", "ARG", "ASN", "ASP", "CYS", "GLN", "GLU", "GLY", "HIS", "ILE",
    "LEU", "LYS", "MET", "PHE", "PRO", "SER", "THR", "TRP", "TYR", "VAL",
  ];
  const SET_AA = new Set(AMINOACIDOS_PADRAO);

  const TRES_PARA_UMA = {
    ALA: "A", ARG: "R", ASN: "N", ASP: "D", CYS: "C", GLN: "Q", GLU: "E",
    GLY: "G", HIS: "H", ILE: "I", LEU: "L", LYS: "K", MET: "M", PHE: "F",
    PRO: "P", SER: "S", THR: "T", TRP: "W", TYR: "Y", VAL: "V",
  };

  // Identificador da cadeia: coluna 22 (indice 21). Vazio -> "_".
  function idCadeia(linha) {
    const c = linha.charAt(21);
    return c === " " || c === "" ? "_" : c;
  }

  function ehRegistro(linha, tipos) {
    const reg = linha.substring(0, 6).trim();
    return tipos.includes(reg);
  }

  function listarCadeias(texto) {
    const vistos = [];
    for (const linha of texto.split("\n")) {
      if (linha.length >= 22 && ehRegistro(linha, ["ATOM", "HETATM", "TER"])) {
        const cid = idCadeia(linha);
        if (!vistos.includes(cid)) vistos.push(cid);
      }
    }
    return vistos;
  }

  function separarCadeias(texto) {
    const cadeias = {};
    for (const linha of texto.split("\n")) {
      if (linha.length >= 22 && ehRegistro(linha, ["ATOM", "HETATM", "TER", "ANISOU"])) {
        const cid = idCadeia(linha);
        (cadeias[cid] = cadeias[cid] || []).push(linha);
      }
    }
    const resultado = {};
    for (const cid of Object.keys(cadeias).sort()) {
      resultado[cid] = cadeias[cid].join("\n") + "\nEND\n";
    }
    return resultado;
  }

  // Lista [nomeResiduo, numResiduo] de uma cadeia, na ordem, sem repetir.
  function residuosOrdenados(texto, cidAlvo) {
    const seq = [];
    const vistos = new Set();
    for (const linha of texto.split("\n")) {
      if (linha.length < 27 || !linha.startsWith("ATOM")) continue;
      const cid = idCadeia(linha);
      if (cidAlvo != null && cid !== cidAlvo) continue;
      const nome = linha.substring(17, 20).trim();
      const num = linha.substring(22, 27).trim();
      const chave = cid + "|" + nome + "|" + num;
      if (!vistos.has(chave)) {
        vistos.add(chave);
        seq.push([nome, num]);
      }
    }
    return seq;
  }

  function estatisticas(texto) {
    const porCadeia = {};
    for (const linha of texto.split("\n")) {
      if (linha.length < 27 || !linha.startsWith("ATOM")) continue;
      const cid = idCadeia(linha);
      const nome = linha.substring(17, 20).trim();
      const num = linha.substring(22, 27).trim();
      const d = porCadeia[cid] || (porCadeia[cid] = { atomos: 0, residuos: new Set(), aa: new Set() });
      d.atomos += 1;
      d.residuos.add(num);
      if (SET_AA.has(nome)) d.aa.add(nome);
    }
    const cadeias = Object.keys(porCadeia).sort().map((cid) => ({
      cadeia: cid,
      atomos: porCadeia[cid].atomos,
      residuos: porCadeia[cid].residuos.size,
      aminoacidos_diferentes: porCadeia[cid].aa.size,
    }));
    let maior = null, menor = null;
    if (cadeias.length) {
      maior = cadeias.reduce((a, b) => (b.residuos > a.residuos ? b : a));
      menor = cadeias.reduce((a, b) => (b.residuos < a.residuos ? b : a));
    }
    return { cadeias, maior, menor };
  }

  function frequenciaAminoacidos(texto, cidAlvo) {
    const contagem = {};
    for (const aa of AMINOACIDOS_PADRAO) contagem[aa] = 0;
    for (const [nome] of residuosOrdenados(texto, cidAlvo)) {
      if (contagem.hasOwnProperty(nome)) contagem[nome] += 1;
    }
    const total = Object.values(contagem).reduce((s, v) => s + v, 0);
    const itens = AMINOACIDOS_PADRAO.map((aa) => ({
      aa,
      contagem: contagem[aa],
      porcentagem: total ? Math.round((10000 * contagem[aa]) / total) / 100 : 0,
    }));
    return { total, itens };
  }

  function sequenciaUmaLetra(texto, cid) {
    return residuosOrdenados(texto, cid)
      .map(([nome]) => TRES_PARA_UMA[nome] || "X")
      .join("");
  }

  function buscarMotivo(texto, motivo) {
    motivo = motivo.map((m) => m.trim().toUpperCase()).filter(Boolean);
    const ocorrencias = [];
    if (!motivo.length) return ocorrencias;
    const n = motivo.length;
    for (const cid of listarCadeias(texto)) {
      const seq = residuosOrdenados(texto, cid);
      const nomes = seq.map((r) => r[0]);
      for (let i = 0; i <= nomes.length - n; i++) {
        let bate = true;
        for (let k = 0; k < n; k++) if (nomes[i + k] !== motivo[k]) { bate = false; break; }
        if (bate) {
          ocorrencias.push({
            cadeia: cid,
            posicao: seq[i][1],
            residuos: Array.from({ length: n }, (_, k) => seq[i + k][1]),
          });
        }
      }
    }
    return ocorrencias;
  }

  function analisarCompleto(texto) {
    const cadeias = listarCadeias(texto);
    const frequencias = {};
    const sequencias = {};
    for (const cid of cadeias) {
      frequencias[cid] = frequenciaAminoacidos(texto, cid);
      sequencias[cid] = sequenciaUmaLetra(texto, cid);
    }
    return {
      cadeias,
      estatisticas: estatisticas(texto),
      frequencias,
      sequencias,
    };
  }

  // Extrai metadados dos registros de cabeçalho do PDB (HEADER, TITLE, COMPND,
  // SOURCE, EXPDTA, REMARK 2, HET/HETNAM). Funciona para upload e para ID.
  function extrairMetadados(texto) {
    const meta = { id: null, classificacao: null, data: null, titulo: null,
      metodo: null, resolucao: null, organismos: [], moleculas: [], ligantes: [] };
    const titulo = [], hetCount = {}, hetNome = {};
    for (const l of texto.split("\n")) {
      const rec = l.substring(0, 6).trim();
      if (rec === "HEADER") {
        meta.classificacao = l.substring(10, 50).trim() || null;
        meta.data = l.substring(50, 59).trim() || null;
        meta.id = l.substring(62, 66).trim() || null;
      } else if (rec === "TITLE") {
        titulo.push(l.substring(10).trim());
      } else if (rec === "EXPDTA") {
        meta.metodo = ((meta.metodo ? meta.metodo + " " : "") + l.substring(10).trim()).trim();
      } else if (l.startsWith("REMARK   2 RESOLUTION")) {
        const m = l.match(/([\d.]+)\s*ANGSTROM/);
        if (m) meta.resolucao = parseFloat(m[1]);
      } else if (rec === "COMPND") {
        const m = l.match(/MOLECULE:\s*(.+?);?\s*$/);
        if (m && !meta.moleculas.includes(m[1].trim())) meta.moleculas.push(m[1].trim());
      } else if (rec === "SOURCE") {
        const m = l.match(/ORGANISM_SCIENTIFIC:\s*(.+?);?\s*$/);
        if (m && !meta.organismos.includes(m[1].trim())) meta.organismos.push(m[1].trim());
      } else if (rec === "HET") {
        const id = l.substring(7, 10).trim();
        if (id && id !== "HOH") hetCount[id] = (hetCount[id] || 0) + 1;
      } else if (rec === "HETNAM") {
        const id = l.substring(11, 14).trim();
        if (id) hetNome[id] = ((hetNome[id] ? hetNome[id] + " " : "") + l.substring(15).trim()).trim();
      }
    }
    meta.titulo = titulo.join(" ").replace(/\s+/g, " ").trim() || null;
    meta.ligantes = Object.keys(hetCount).map((id) => ({ id, count: hetCount[id], nome: hetNome[id] || null }));
    return meta;
  }

  function validar(texto) {
    if (!texto.includes("ATOM") && !texto.includes("HETATM")) {
      throw new Error("Arquivo invalido: nenhuma linha ATOM/HETATM encontrada.");
    }
  }

  return {
    AMINOACIDOS_PADRAO,
    listarCadeias, separarCadeias, estatisticas, frequenciaAminoacidos,
    sequenciaUmaLetra, buscarMotivo, analisarCompleto, validar, extrairMetadados,
  };
})();
